### ISSUE TITLE
Video output is slanted / skewed

### ISSUE THREAD

**Issue by ryandesign** [TomHarte/CLK#1173]
On my machine (a 2012 15" MacBook Pro with Retina Display running macOS Monterey 12.6.8) Clock Signal (currently version 2023-09-10 but it has been the same in all versions I tried) has slanted or skewed output. Lines which should be horizontal aren't; they are slightly diagonal. It happens in both Apple II and Macintosh emulation; I haven't tried others.

![appleii](https://github.com/TomHarte/CLK/assets/429209/8e6c453f-3d50-4e98-97e4-a031f57f7c37)
![macintosh](https://github.com/TomHarte/CLK/assets/429209/83cd7da0-dc8b-4a5a-a710-a2c17da7962b)

**Comment by ryandesign** [comment]
I also saw this problem on a 2016 15" MacBook Pro with Touch Bar running macOS Sonoma 14.0. (The screenshots in #1181 were taken on that machine.) However, it must not affect all machines, since the screenshots in the README do not exhibit this problem.

I tried to reproduce the problem on a late-2011 13" MacBook Pro but it cannot run Clock Signal because it does not support Metal.

**Comment by ryandesign** [comment]
The SDL/OpenGL version doesn't have this problem running on the same Mac.

**Comment by TomHarte** [comment]
Yeah, this is a design issue; the CRT emulation because it is really, really emulating a flying dot, actually generates a slightly lower coordinate on the right than on the left.

The OpenGL outputter internally applies a reverse skew that approximately levels them out (as in: hopefully to within a whole subsampled pixel), with the 'approximately' bit being because e.g. an Atari 2600 cartridge that generates a full display of 259 lines is going to produce scans that are slightly more slanted than one which generates 263.

Will need to rethink where proper responsibility should lie for this; I guess the scan target should be able to indicate that it needs truly-horizontal scans and the CRT should deal with it.

**Comment by ryandesign** [comment]
> Yeah, this is a design issue; the CRT emulation because it is really, really emulating a flying dot, actually generates a slightly lower coordinate on the right than on the left.

Are you saying real CRTs have slanted lines? I've never noticed that, though the way that Clock Signal has to draw its slanted lines on an existing operating system's straight lines of pixels certainly would make it much more noticeable. 

> The OpenGL outputter internally applies a reverse skew that approximately levels them out (as in: hopefully to within a whole subsampled pixel), with the 'approximately' bit being because e.g. an Atari 2600 cartridge that generates a full display of 259 lines is going to produce scans that are slightly more slanted than one which generates 263.

I admit zero familiarity with the Atari 2600, real or emulated. I do see now that different cartridges used different numbers of scanlines and that TVs were apparently fine with that. But why does different numbers of scanlines require that they be slanted?

> Will need to rethink where proper responsibility should lie for this; I guess the scan target should be able to indicate that it needs truly-horizontal scans and the CRT should deal with it.

Under what circumstances would you actually want non-horizontal scans?
